### ISSUE TITLE
TLS validation error with 0.6.0, it worked without issues in 0.5.4

### ISSUE THREAD

**Issue by simonswine** [projectcalico/kube-controllers#102]
## Expected Behavior

The cert validation should respects SANs of cert like it did with 0.5.4

## Current Behavior

After upgrading to 0.6.0 policy controller I am seeing this error:

```
2017-06-25 10:32:36,538 5 INFO Restarting watch on resource: Pod
/tmp/_MEIfdJBde/requests/packages/urllib3/connection.py:340: SubjectAltNameWarning: Certificate for kubernetes.default has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/shazow/urllib3/issues/497 for details.)
2017-06-25 10:32:36,539 5 WARNING Connection error: NetworkPolicy ...hostname 'kubernetes.default' doesn't match u'system:kube-apiserver'
```
Seems like there is some problem for urllib3 to check the SANs. Everything is working fine switching back to 0.5.4

Worth noting I am using a cert from vault that looks like that (and worked fine before):

```json
{
  "subject": {
    "common_name": "system:kube-apiserver",
    "names": [
      "system:kube-apiserver"
    ]
  },
  "issuer": {
    "common_name": "Kubernetes nonprod-devcluster/k8s CA",
    "names": [
      "Kubernetes nonprod-devcluster/k8s CA"
    ]
  },
  "serial_number": "541755765460935537011784003202274224647913608447",
  "sans": [
    "system:kube-apiserver",
    "kubernetes",
    "kubernetes.default",
    "kubernetes.default.svc",
    "kubernetes.default.svc.cluster.local",
    "localhost",
    "10.99.22.244",
    "10.254.0.1",
    "127.0.0.1"
  ],
  "not_before": "2017-06-25T09:35:04Z",
  "not_after": "2017-07-02T09:35:34Z",
  "sigalg": "SHA256WithRSA",
  "authority_key_id": "58:40:31:93:45:89:F2:3B:A8:CF:CE:BE:F6:E2:D9:EB:DD:80:1C:B3",
  "subject_key_id": "F2:1E:BF:62:54:5E:B3:35:C3:77:AA:91:AB:D4:20:8A:61:56:E4:D1",
  "pem": "-----BEGIN CERTIFICATE-----\nMIIEITCCAwmgAwIBAgIUXuUrMI/UZZsMVYs7trv2ckBUFP8wDQYJKoZIhvcNAQEL\nBQAwLzEtMCsGA1UEAxMkS3ViZXJuZXRlcyBub25wcm9kLWRldmNsdXN0ZXIvazhz\nIENBMB4XDTE3MDYyNTA5MzUwNFoXDTE3MDcwMjA5MzUzNFowIDEeMBwGA1UEAxMV\nc3lzdGVtOmt1YmUtYXBpc2VydmVyMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIB\nCgKCAQEAtDLO7D67iBMGclreUOhSsB3NgNWt6yKS7WAnI0VuyqOUXuhwUVZpdMmv\nfp34PwmzIRhp2eI6JcIoDY2PpOqXrdK2k2TTXnbC3ntiiozOMBNpY8+XgHjlWSht\nA/FSj1JpThuAS4B1Wp/xVvQBSnGuV3EofnZhkf5zGcbycuDULAp6TgW+SbAcnJxp\nXOgENk4/7HCTzAGz0E0TxYNFeCxHYQqGZftGn0i6V1YFhHDNPspkmKMAe+M8bZl9\np0WVl4TYPh5zLsaShjaIJrMTUyu9qShyy+AKrukT858qVehrBROW1Ax7dugA5NSP\nJp/jrvVOLDBoV5wYN761Lmrscbrq/wIDAQABo4IBQjCCAT4wDgYDVR0PAQH/BAQD\nAgOoMBMGA1UdJQQMMAoGCCsGAQUFBwMBMB0GA1UdDgQWBBTyHr9iVF6zNcN3qpGr\n1CCKYVbk0TAfBgNVHSMEGDAWgBRYQDGTRYnyO6jPzr724tnr3YAcszCB1gYDVR0R\nBIHOMIHLghVzeXN0ZW06a3ViZS1hcGlzZXJ2ZXKCN2FwaS5ub25wcm9kLWRldmNs\ndXN0ZXIubm9ucHJvZC1wcml2YXRlLnA5cy5qZXRzdGFjay5uZXSCCmt1YmVybmV0\nZXOCEmt1YmVybmV0ZXMuZGVmYXVsdIIWa3ViZXJuZXRlcy5kZWZhdWx0LnN2Y4Ik\na3ViZXJuZXRlcy5kZWZhdWx0LnN2Yy5jbHVzdGVyLmxvY2Fsgglsb2NhbGhvc3SH\nBApjFvSHBAr+AAGHBH8AAAEwDQYJKoZIhvcNAQELBQADggEBABac9aVwPdus91mX\n76a3O6eTB0vCuLpyf7yiA45IJK/ARu5msNGn0a4x35XSt/d/rZJjn/2QTtAb+rpw\nQik+LBL2nc/7WW8igWs5z+40PIHJheWSrFQnwttQSzI2Jjzb1TdsOxhDxU2rwYyf\nfz/XiJhqEORo43uOCFtxsWH+r1EERRZPZpOIzCbtBHX3ITqpE/oUZLz1/ryyquHi\nGP6xkL8YpY/Ey4OKlRF54e4v1v2MVmi3lQm1uCfABWB9Xf6r6BTQ/CfykMXiruuo\n8mfw5C7qMf/VKqYmzQ4S2MmqQddsS3pIpBN+F/iz9GPGV/+m87sdRfkQbXrQBiKN\nlNOO73E=\n-----END CERTIFICATE-----\n"
}

```

**Comment by tmjd** [comment]
I haven't used vault before so please forgive the question but in the json you included, does that "pem" get get written as the cert for the policy-controller?  If not, could you please grab a copy of the cert that is given to the policy-controller.

I took a look at that one in that "pem" key and as best as I can tell shouldn't cause a problem. I found an issue with urllib3 with older versions of python 2.7.2 but from looking at the build logs for the container we are using 2.7.12 (see https://github.com/shazow/urllib3/issues/523).  I found that issue when looking for similar errors with urllib3.

@caseydavenport Any thoughts here?

**Comment by simonswine** [comment]
@tmjd the problem is within the part of policy-controller that validates the kubernetes API server certificate. (My output here is just a cfssl output of the API server certificate). It works with 0.5.4, so the cert is not the issue (It also works with all other api clients on the cluster). I am not too sure how the changes that occurred between the two versions could have affected this: https://github.com/projectcalico/k8s-policy/compare/v0.5.4...v0.6.0

**Comment by tmjd** [comment]
I understand that the cert works fine with v0.5.4.  I was just trying to reproduce the problem with the cert you are using to identify the component that is causing the issue.

**Comment by tmjd** [comment]
Sorry I see where I went wrong in my original question about the certificate.  I understand that the certificate that is causing the problem for the policy-controller is the one that the kube-apiserver is presenting in the connection.  So my question should have been does that "pem" get get written as the cert for the kube-apiserver?  If so then I do not understand the error message because it clearly has a "Subject Alternatve Name" unless it is a problem to have `DNS:system:kube-apiserver` which would not be a valid DNS name (due to the `:`, I believe).

**Comment by bcreane** [comment]
recheck with k8s-policy/master

**Comment by caseydavenport** [comment]
The latest master / RC releases no longer use urllib3 so I don't think will encounter this issue.

**Comment by bcreane** [comment]
@simonswine - please recheck this issue against v2.6.0 (which should be published in the next day or so). Thanks!

**Comment by heschlie** [comment]
@simonswine were you able to recheck this against Calico v2.6? We've made significant changes in this area and are no longer using urllib3.

**Comment by simonswine** [comment]
Thanks for the reminder (and obviously the fix), it works from 0.7.x onwards for me
